commit 99780da7209605bf9f226eac3eac30ab2bc9427c
Author: Fujii Masao <fujii@postgresql.org>
Date:   Thu Nov 20 09:14:02 2025 +0900

    Add HINT listing valid encodings to encode() and decode() errors.
    
    This commit updates encode() and decode() so that when an invalid encoding
    is specified, their error message includes a HINT listing all valid encodings.
    This helps users quickly see which encodings are supported without needing
    to consult the documentation.
    
    Author: Shinya Sugamoto <shinya34892@gmail.com>
    Reviewed-by: Chao Li <lic@highgo.com>
    Reviewed-by: Daniel Gustafsson <daniel@yesql.se>
    Reviewed-by: Masahiko Sawada <sawada.mshk@gmail.com>
    Reviewed-by: Peter Eisentraut <peter@eisentraut.org>
    Reviewed-by: Nathan Bossart <nathandbossart@gmail.com>
    Reviewed-by: Fujii Masao <masao.fujii@gmail.com>
    Discussion: https://postgr.es/m/CAAe3y+99sfPv8UDF1VM-rC1i5HBdqxUh=2HrbJJFm2+i=1OwOw@mail.gmail.com

diff --git a/src/backend/utils/adt/encode.c b/src/backend/utils/adt/encode.c
index aabe9913eee..c813ee1258b 100644
--- a/src/backend/utils/adt/encode.c
+++ b/src/backend/utils/adt/encode.c
@@ -62,11 +62,13 @@ binary_encode(PG_FUNCTION_ARGS)
 
 	enc = pg_find_encoding(namebuf);
 	if (enc == NULL)
 		ereport(ERROR,
 				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
-				 errmsg("unrecognized encoding: \"%s\"", namebuf)));
+				 errmsg("unrecognized encoding: \"%s\"", namebuf),
+				 errhint("Valid encodings are \"%s\", \"%s\", \"%s\", and \"%s\".",
+						 "base64", "base64url", "escape", "hex")));
 
 	dataptr = VARDATA_ANY(data);
 	datalen = VARSIZE_ANY_EXHDR(data);
 
 	resultlen = enc->encode_len(dataptr, datalen);
@@ -110,11 +112,13 @@ binary_decode(PG_FUNCTION_ARGS)
 
 	enc = pg_find_encoding(namebuf);
 	if (enc == NULL)
 		ereport(ERROR,
 				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
-				 errmsg("unrecognized encoding: \"%s\"", namebuf)));
+				 errmsg("unrecognized encoding: \"%s\"", namebuf),
+				 errhint("Valid encodings are \"%s\", \"%s\", \"%s\", and \"%s\".",
+						 "base64", "base64url", "escape", "hex")));
 
 	dataptr = VARDATA_ANY(data);
 	datalen = VARSIZE_ANY_EXHDR(data);
 
 	resultlen = enc->decode_len(dataptr, datalen);
diff --git a/src/test/regress/expected/strings.out b/src/test/regress/expected/strings.out
index b9dc08d5f61..727304f60e7 100644
--- a/src/test/regress/expected/strings.out
+++ b/src/test/regress/expected/strings.out
@@ -2573,10 +2573,17 @@ SELECT decode(encode('\x1234567890abcdef00', 'escape'), 'escape');
         decode        
 ----------------------
  \x1234567890abcdef00
 (1 row)
 
+-- report an error with a hint listing valid encodings when an invalid encoding is specified
+SELECT encode('\x01'::bytea, 'invalid');  -- error
+ERROR:  unrecognized encoding: "invalid"
+HINT:  Valid encodings are "base64", "base64url", "escape", and "hex".
+SELECT decode('00', 'invalid');           -- error
+ERROR:  unrecognized encoding: "invalid"
+HINT:  Valid encodings are "base64", "base64url", "escape", and "hex".
 --
 -- base64url encoding/decoding
 --
 SET bytea_output TO hex;
 -- Simple encoding/decoding
diff --git a/src/test/regress/sql/strings.sql b/src/test/regress/sql/strings.sql
index a2a91523404..88aa4c2983b 100644
--- a/src/test/regress/sql/strings.sql
+++ b/src/test/regress/sql/strings.sql
@@ -813,10 +813,14 @@ SELECT encode(('\x' || repeat('1234567890abcdef0001', 7))::bytea, 'base64');
 SELECT decode(encode(('\x' || repeat('1234567890abcdef0001', 7))::bytea,
                      'base64'), 'base64');
 SELECT encode('\x1234567890abcdef00', 'escape');
 SELECT decode(encode('\x1234567890abcdef00', 'escape'), 'escape');
 
+-- report an error with a hint listing valid encodings when an invalid encoding is specified
+SELECT encode('\x01'::bytea, 'invalid');  -- error
+SELECT decode('00', 'invalid');           -- error
+
 --
 -- base64url encoding/decoding
 --
 SET bytea_output TO hex;
 
